osquery/events/darwin/endpointsecurity.cpp

```diff
@@ -0,0 +1,220 @@
+
+Status EndpointSecurityPublisher::setUp() {
+  if (__builtin_available(macos 10.15, *)) {
+    if (FLAGS_disable_endpointsecurity) {
+      return Status::failure(1, "ES disabled via configuration");
+    }
+
+    auto handler = ^(es_client_t* client, const es_message_t* message) {
+      handleMessage(message);
+    };
+
+    auto result = es_new_client(&es_client_, handler);
+    switch (result) {
+    case ES_NEW_CLIENT_RESULT_SUCCESS: {
+      es_client_success_ = true;
+      auto ver = SQL::selectAllFrom("os_version");
+      // check if macOS version is 10.15.4 or higher
+      if (ver.front().at("major") == "10" && ver.front().at("minor") == "15" &&
+          std::stoi(ver.front().at("patch")) >= 4) {
+        macos_15_4_higher_ = true;
+      }
+    }
+      return Status::success();
+    case ES_NEW_CLIENT_RESULT_ERR_INVALID_ARGUMENT:
+      return Status::failure(
+          1, "Invalid args provided to EndpointSecurity client");
+    case ES_NEW_CLIENT_RESULT_ERR_INTERNAL:
+      return Status::failure(
+          1, "EndpointSecurity client cannot communicate with ES subsystem");
+    case ES_NEW_CLIENT_RESULT_ERR_NOT_ENTITLED:
+      return Status::failure(1, "EndpointSecurity client lacks entitlement");
+    case ES_NEW_CLIENT_RESULT_ERR_NOT_PERMITTED:
+      return Status::failure(
+          1, "EndpointSecurity client lacks user TCC permissions");
+    case ES_NEW_CLIENT_RESULT_ERR_NOT_PRIVILEGED:
+      return Status::failure(1,
+                             "EndpointSecurity client is not running as root");
+    case ES_NEW_CLIENT_RESULT_ERR_TOO_MANY_CLIENTS:
+      return Status::failure(
+          1, "Too many EndpointSecurity clients running on system");
+    default:
+      return Status::failure(1, "EndpointSecurity: Unknown Error");
+    }
+  } else {
+    return Status::failure(1,
+                           "EndpointSecurity is only available on macOS 10.15");
+  }
+}
+
+void EndpointSecurityPublisher::configure() {
+  if (es_client_ == nullptr) {
+    return;
+  }
+
+  auto cache = es_clear_cache(es_client_);
+  if (cache != ES_CLEAR_CACHE_RESULT_SUCCESS) {
+    VLOG(1) << "Couldn't clear cache for EndpointSecurity client";
+    return;
+  }
+
+  for (auto& sub : subscriptions_) {
+    auto sc = getSubscriptionContext(sub->context);
+    auto events = sc->es_event_subscriptions_;
+    auto es_sub = es_subscribe(es_client_, &events[0], events.size());
+    if (es_sub != ES_RETURN_SUCCESS) {
+      VLOG(1) << "Couldn't subscribe to ES subsystem";
+    }
+  }
+}
+
+void EndpointSecurityPublisher::tearDown() {
+  if (es_client_ == nullptr) {
+    return;
+  }
+  es_unsubscribe_all(es_client_);
+
+  // calling es_delete_client on macOS 10.15.0 - 10.15.3 leads to a crash
+  if (es_client_success_ && macos_15_4_higher_) {
+    es_delete_client(es_client_);
+  }
+}
+
+static inline std::string getPath(const es_process_t* p) {
+  return p->executable->path.length > 0 ? p->executable->path.data : "";
```

Comment:
The docs say that the length does not include the null terminator.  However, the data strings should be null-terminated.  So we could access path, signing_id, team_id without these getters?